File: sw/JtagTap.cpp
#include <JtagTap.hpp>

#include <array>
#include <stdexcept>

namespace {
    std::array<const std::string, 16> stateNames({
        "TestLogicReset", "RunTestIdle",
        "SelectDRScan", "CaptureDR", "ShiftDR", "Exit1DR", "PauseDR", "Exit2DR", "UpdateDR",
        "SelectIRScan", "CaptureIR", "ShiftIR", "Exit1IR", "PauseIR", "Exit2IR", "UpdateIR"
    });
}

JtagTap::JtagTap(unsigned irLen)
: irLen_(irLen)
{
  if ( irLen > sizeof(ir_)*8 ) {
    throw std::runtime_error("irLen too big");
  }
  ir_    = BYPASS;
  state_ = State::TestLogicReset;
}

const std::string &
JtagTap::toString(State state)
{
	return stateNames[static_cast<unsigned>(state)];
}

size_t
JtagTap::getDRLen()
{
    size_t sz = dri_.size();
    // also correct if sz == 0 (drRemBits initialized to 8)
    return 8*sz - (8-drRemBits_);
}

JtagTap::State
JtagTap::nextState(bool tms, bool tdi, bool tdo)
{
    switch ( state_ ) {
       case State::TestLogicReset:
           ir_ = BYPASS;
           if ( ! tms ) {
              state_ = State::RunTestIdle;
           }
           break;
       case State::RunTestIdle:
           if ( tms ) {
              state_ = State::SelectDRScan;
           }
           break;
       case State::SelectDRScan:
           if ( tms ) {
              state_ = State::SelectIRScan;
           } else {
              state_ = State::CaptureDR;
           }
           break;
       case State::CaptureDR:
           dri_.clear();
           dro_.clear();
           drRemBits_ = 8;
           if ( tms ) {
              state_ = State::Exit1DR;
           } else {
              state_ = State::ShiftDR;
           }
           break;
       case State::ShiftDR:
           if ( drRemBits_ >= sizeof(uint8_t)*8 ) {
             dri_.push_back(0x00);
             dro_.push_back(0x00);
             drRemBits_ = 0;
           }
           if ( tdi ) {
               dri_[dri_.size() - 1] |= (1<<drRemBits_);
           }
           if ( tdo ) {
               dro_[dro_.size() - 1] |= (1<<drRemBits_);
           }
           ++drRemBits_;
           
           if ( tms ) {
              state_ = State::Exit1DR;
           }
           break;
       case State::Exit1DR:
           if ( tms ) {
              state_ = State::UpdateDR;
           } else {
              state_ = State::PauseDR;
           }
           break;
       case State::PauseDR:
           if ( tms ) {
              state_ = State::Exit2DR;
           }
           break;
       case State::Exit2DR:
           if ( tms ) {
              state_ = State::UpdateDR;
           } else {
              state_ = State::ShiftDR;
           }
           break;
       case State::UpdateDR:
           updateDR(dri_, dro_, drRemBits_);
           if ( tms ) {
              state_ = State::SelectDRScan;
           } else {
              state_ = State::RunTestIdle;
           }
           break;
       case State::SelectIRScan:
           if ( tms ) {
              state_ = State::TestLogicReset;
           } else {
              state_ = State::CaptureIR;
           }
           break;
       case State::CaptureIR:
	   irLen_ = 0;
	   ir_    = 0;
           if ( tms ) {
              state_ = State::Exit1IR;
           } else {
              state_ = State::ShiftIR;
           }
           break;
       case State::ShiftIR:
	   if ( irLen_ >= sizeof(ir_)*8 ) {
               throw std::runtime_error("irLen too big");
	   }
	   if ( tdi ) {
	       ir_ |= (1ULL<<irLen_);
	   }
	   ++irLen_;
           if ( tms ) {
              state_ = State::Exit1IR;
           }
           break;
       case State::Exit1IR:
           if ( tms ) {
              state_ = State::UpdateIR;
           } else {
              state_ = State::PauseIR;
           }
           break;
       case State::PauseIR:
           if ( tms ) {
              state_ = State::Exit2IR;
           }
           break;
       case State::Exit2IR:
           if ( tms ) {
              state_ = State::UpdateIR;
           } else {
              state_ = State::ShiftIR;
           }
           break;
       case State::UpdateIR:
	   updateIR(ir_, irLen_);
           if ( tms ) {
              state_ = State::SelectDRScan;
           } else {
              state_ = State::RunTestIdle;
           }
           break;
    }
    return state_;
}


void
JtagTap::printDRi(FILE *f)
{
	printdr(f, true);
}
	
void
JtagTap::printDRo(FILE *f)
{
	printdr(f, false);
}

void
JtagTap::printdr(FILE *f, bool dri)
{
	const std::vector<uint8_t> &d = dri ? dri_ : dro_;
	fprintf(f, "MSB has %d bits\n", drRemBits_);
	fprintf(f, "0x");
	for (ssize_t i = d.size() - 1; i >= 0; --i) {
		fprintf(f, "%02x", d[i]);
	}
	fprintf(f,"\n");
}

File: py/jtagtap.py
from enum import Enum
class JtagTap:
  class State(Enum):
    TEST_LOGIC_RESET = 0
    RUN_TEST_IDLE    = 1
    SELECT_DR_SCAN   = 2
    CAPTURE_DR       = 3
    SHIFT_DR         = 4
    EXIT1_DR         = 5
    PAUSE_DR         = 6
    EXIT2_DR         = 7
    UPDATE_DR        = 8
    SELECT_IR_SCAN   = 9
    CAPTURE_IR       = 10
    SHIFT_IR         = 11
    EXIT1_IR         = 12
    PAUSE_IR         = 13
    EXIT2_IR         = 14
    UPDATE_IR        = 15

  def __init__(self, ir_len):
    self.state  = JtagTap.State.TEST_LOGIC_RESET
    self.ir_len = ir_len
    self.byp    = ((1<<self.ir_len) - 1)
    self.ir     = self.byp
    self.sr     = 0
  

  def state(self):
    return self.state

  def tmsSeq(self, seq):
    for i in seq:
      self.nextState(i)

  def nextState(self, tms, tdi=False):
    if   ( self.state == self.State.TEST_LOGIC_RESET):
      if ( not  tms ):
        self.state = self.State.RUN_TEST_IDLE
      self.ir  = self.byp
    elif ( self.state == self.State.RUN_TEST_IDLE   ):
      if ( tms ):
        self.state = self.State.SELECT_DR_SCAN
    elif ( self.state == self.State.SELECT_DR_SCAN  ):
      if ( tms ):
        self.state = self.State.SELECT_IR_SCAN
      else:
        self.state = self.State.CAPTURE_DR
    elif ( self.state == self.State.CAPTURE_DR      ):
      if ( tms ):
        self.state = self.State.EXIT1_DR
      else:
        self.state = self.State.SHIFT_DR
    elif ( self.state == self.State.SHIFT_DR        ):
      if ( tms ):
        self.state = self.State.EXIT1_DR
    elif ( self.state == self.State.EXIT1_DR        ):
      if ( tms ):
        self.state = self.State.UPDATE_DR
      else:
        self.state = self.State.PAUSE_DR
    elif ( self.state == self.State.PAUSE_DR        ):
      if ( tms ):
        self.state = self.State.EXIT2_DR
    elif ( self.state == self.State.EXIT2_DR        ):
      if ( tms ):
        self.state = self.State.UPDATE_DR
      else:
        self.state = self.State.SHIFT_DR
    elif ( self.state == self.State.UPDATE_DR       ):
      if ( tms ):
        self.state = self.State.SELECT_DR_SCAN
      else:
        self.state = self.State.RUN_TEST_IDLE
    elif ( self.state == self.State.SELECT_IR_SCAN  ):
      if ( tms ):
        self.state = self.State.TEST_LOGIC_RESET
      else:
        self.state = self.State.CAPTURE_IR
    elif ( self.state == self.State.CAPTURE_IR      ):
      if ( tms ):
        self.state = self.State.EXIT1_IR
      else:
        self.state = self.State.SHIFT_IR
      self.sr = 1;
    elif ( self.state == self.State.SHIFT_IR        ):
      if ( tms ):
        self.state = self.State.EXIT1_IR
      if ( tdi ):
         self.sr |= (1<<self.ir_len)
      self.sr >>=1
    elif ( self.state == self.State.EXIT1_IR        ):
      if ( tms ):
        self.state = self.State.UPDATE_IR
      else:
        self.state = self.State.PAUSE_IR
    elif ( self.state == self.State.PAUSE_IR        ):
      if ( tms ):
        self.state = self.State.EXIT2_IR
    elif ( self.state == self.State.EXIT2_IR        ):
      if ( tms ):
        self.state = self.State.UPDATE_IR
      else:
        self.state = self.State.SHIFT_IR
    elif ( self.state == self.State.UPDATE_IR       ):
      if ( tms ):
        self.state = self.State.SELECT_DR_SCAN
      else:
        self.state = self.State.RUN_TEST_IDLE
      self.ir = (self.sr & self.byp)
    else:
      raise RuntimeError("Invalid State")

  def tdo(self):
    if ( self.state != self.State.SHIFT_DR and self.state != self.State.SHIFT_IR ):
      return None
    return self.sr & 1
